### ISSUE TITLE
support for defer: true 

### ISSUE THREAD

**Issue by jaraquistain** [makeomatic/redux-connect#3]
In `redux-async-connect` I was able to utilize a `defer: true` option in order to defer the fetching of the data until after the view has transitioned. After switching to `redux-connect` it seems like that use case is no longer supported. 

Are there plans to add this functionality back in? My specific use case is that when I fetch an external entity I keep its data in my store along with a `loaded` flag and only fetch the data if `loaded` is `false` . In some cases when I transition away from this entity's view I will set its `loaded` flag back to false, but not clear out the data. The result is that if I navigate back to that view I'm able to immediately display previously loaded data while the fresh data is fetched from the server, resulting in a much faster perceived page speed.

Admittedly, I'm not sure how much of that logic was built into `redux-async-connect` and how much of it was manually implemented via the [react-redux-universal-hot-example](https://github.com/erikras/react-redux-universal-hot-example/issues/616) that my app is based off of. All I can say for sure is that when switching to `redux-connect` if you navigate to a page with `deferred: true` then the data loads fine if its coming from the server, but if its a page transition in the client the data never loads.

**Comment by AVVS** [comment]
There are 2 things for that:

1. https://github.com/makeomatic/redux-connect/blob/master/modules/components/AsyncConnect.js#L35-L42

Here we have a check that is performed on whether data was loaded or not

2. https://github.com/makeomatic/redux-connect/blob/master/modules/components/AsyncConnect.js#L58

If you pass `filter` props to ReduxAsyncComponents - this is where you can add logic to whether you need to load it or not (defer example) - it should be different on server & client.

As for data not loading on client - there is pretty simply boolean logic here, it is either completely loaded or not and thats what redux-connect is tracking. If we need to do partial checks - then we'll have to understand when to set data as not loaded. For now load is always triggered after routing - and it's ultimately your responsibility to avoid fetching same data twice. You can incorporate those checks into your own actions in redux, for instance

**Comment by jaraquistain** [comment]
TL;DR: I figured out that if your `promise` property returns a promise then the router will not transition until all the promises have been resolved, whereas if your `promise` does not return anything (or something that isn't an unresolved promise) then the next route is rendered immediately, even while the fetches are still loading (which is exactly what I wanted). Given the way `redux-connect` works I don't currently want to pass a `filter` into `<ReduxAsyncConnect />` because I do want the data to load in all cases. I don't know why `redux-async-connect` would still load things on the client that were filtered out.

EG:

I've got a `/networks/:id` route that renders a `Network` container. This `Network` container requires me to externally fetch a list of `NETWORK_SUBDOMAN` data. The container also indicates whether or not any of the data is currently loading via a spinner. For the sake of argument lets say I've also got a `/home` page that contains a link pointing to `/networks/123`. If my `mapAsyncStateToProps` is this:
```javascript
const asyncStateToProps = NETWORK_SUBDOMAINS.map(subdomain => ({
  promise: ({ store: { dispatch, getState }, params }) => {
    const networkUrl = urlFromId(params.id);
    return !isLoaded(getState(), networkUrl, subdomain) ? dispatch(load(networkUrl, subdomain)) : Promise.resolve();
  }
}));
```
Then when I click my link on `/home` the page will sit until ALL of the data for `/networks/123` is loaded, THEN transition to the view. When `/networks/123` is visible there is no spinner because all the data has finished loading

However, if my `mapAsyncStateToProps` is this:
```javascript
const asyncStateToProps = NETWORK_SUBDOMAINS.map(subdomain => ({
  promise: ({ store: { dispatch, getState }, params }) => {
    const networkUrl = urlFromId(params.id);
    !isLoaded(getState(), networkUrl, subdomain) ? dispatch(load(networkUrl, subdomain)) : Promise.resolve();
  }
}));
```
(the only difference being a lack of a `return` at the bottom)

Then when I click the link on `/home`, the `/networks/123` view is (almost) immediately visible and shows the spinner indicating the data is still loading. When the data is finished loading the spinner disappears and the data replaces it, as expected.

I definitely follow what you're saying and based on that the behavior I'm seeing is exactly as expected. I guess I just don't understand the core differences between `redux-async-connect` and this fork, because `redux-async-connect` behaves entirely differently in the situation I described. 

I was previously passing `filter` to `<ReduxAsyncConnect />` in my `client.js` like this:

```javascript
// Render routes using ReduxAsyncConnect
function renderRouter(props) {
  function filter(item) { return !item.deferred; }
  return <ReduxAsyncConnect { ...props } helpers={{ client }} filter={filter} />;
}

// The configured Router component for the app
const component = (
  <Router render={renderRouter} history={enhancedHistory}>
    {getRoutes(store)}
  </Router>
);
```

and in the case of `redux-connect` I can fully understand how filtering out fetches that contain `deferred` would mean that that data is never fetched on the client. What doesn't make sense is why `redux-async-connect` would still fetch things that are filtered out, but after the new route renders. Maybe there's some magic boilerplate code handling that part? I haven't seen anything that would do something like that.

Anyway, thanks for your response it was very helpful

**Comment by jaraquistain** [comment]
Also, I should probably close this since I've figured out the answer to my original question

**Comment by AVVS** [comment]
@jaraquistain in the boilterplate you are using they have methods to short-circuit loading of data if it was already loaded. Basically when you try to fetch data - you first check if it's loaded and if it was - you just return it via Promise.resolve() or just simple return - doesn't really matter since it's wrapper.

**Comment by jaraquistain** [comment]
UPDATE: Slight modification to what I said earlier. 

never returning the actual promise inside of `promise` as I said I was doing earlier actually breaks server side data fetching because the server thinks the fetch is immediatly finished and sends the response. I've modified my `mapAsyncStateToProps` to be the following:

```javascript
const asyncStateToProps = NETWORK_SUBDOMAINS.map(subdomain => ({
  promise: ({ store: { dispatch, getState }, params }) => {
    const networkUrl = urlFromId(params.id);
    const result = !isLoaded(getState(), networkUrl, subdomain) ? dispatch(load(networkUrl, subdomain)) : Promise.resolve();
    if (!__CLIENT__) { return result; }
  }
}));
```

not the prettiest but it actually works until I think of something better.

**Comment by AVVS** [comment]
alter `result` so that instead of Promise.resolve() it would return the actual value, that way it's effectively a short-circuit.

Also whats `isLoaded` code?

**Comment by jaraquistain** [comment]
@AVVS ok I'll try replacing `Promise.resolve()` with the actual values.

`isLoaded` is my function that I call to determine whether or not whatever we want to fetch is already in the current state or if we need to dispatch the action to get it from the server. It looks like this:

```javascript
export function isLoaded(globalState, networkUrl, subdomain) {
  const thisNetwork = globalState.networks[networkUrl] || {};

  return (subdomain ? (thisNetwork[subdomain] || {}) : thisNetwork).loaded;
}
```

**Comment by AVVS** [comment]
@jaraquistain if you supply `key` alongside `promise` - you will be able to check loaded data easier:

use `getState().reduxAsyncConnect.loadState[key]` to check for loading state, it has 3 props:

```js
{
  loading: Boolean,
  loaded: Boolean,
  error: Object,
}
``

Also, the response data would be stored at `getState().reduxAsyncConnect[key]`

Logic for this is here: https://github.com/makeomatic/redux-connect/blob/master/modules/store.js

**Comment by jaraquistain** [comment]
@AVVS Excellent tip, thank you!

**Comment by badtant** [comment]
@jaraquistain I'm trying to do exactly the same thing as you but I don't really follow what I should change :)

I've removed all my "deferred: true" and the filter from the client code "filter={item => !item.deferred}"
Now server rendering works fine but I don't get my spinner when transitioning on the client. Maybe you can point me in the right direction?

My async connect code is like this now:
```
@asyncConnect([{
	promise: ({store}) => {
		if (!isLoaded(store.getState())) {
			return store.dispatch(loadAction());
		}
	}
}])
```

**Comment by AVVS** [comment]
rendering is delayed until all the data from async connect is loaded, so you wont see it. If you need a spinner you should dispatch those actions in componentDidMount or not return a promise from async connect, otherwise it will wait until its done

**Comment by badtant** [comment]
Thanks!
So right now I have this which seems to be working. Is that the way to go?

```
@asyncConnect([{
	promise: ({store}) => {
		const result = !isLoaded(store.getState()) ? store.dispatch(loadAction()) : null;
		if (!__CLIENT__) {
			return result;
		}
	}
}])
```

**Comment by jaraquistain** [comment]
^ That's essentially what I ended up with and it's been working OK for me so far.

Only difference with mine is that my linter enforces a consistent return statement so my promise is:

```javascript
const result = !isLoaded(getState()) ? dispatch(load()) : null;
return __CLIENT__ ? null : result;
```

**Comment by AVVS** [comment]
@badtant that way client wont wait for result of the promise, and, therefore, wont delay routing. So yes, if that's what you want - it is the way to go

**Comment by lakshyaranganath** [comment]
Was just going through this since I too was wondering how to stop it from blocking routing on the browser, just wondering if someone found a cleaner alternative to the solutions mentioned above?

**Comment by jaraquistain** [comment]
@lakshyaranganath I couldn't think of anything better so I've been sticking with what was discussed above

**Comment by lakshyaranganath** [comment]
Ahh alright! Thanks again! :)

Regards,
Lakshya Ranganath
Freelance Web Developer
+91 90352 09244
